resolver/resolver.go

```diff
@@ -67,77 +76,72 @@ func (r *Resolver) ResolveToLastNode(ctx context.Context, fpath path.Path) (cid.
 		return c, nil, nil
 	}
 
-	nd, err := r.DAG.Get(ctx, c)
+	// create a selector to traverse all path segments but only match the last
+	pathSelector, err := pathLeafSelector(p[:len(p)-1])
 	if err != nil {
 		return cid.Cid{}, nil, err
 	}
 
-	for len(p) > 0 {
-		lnk, rest, err := r.ResolveOnce(ctx, r.DAG, nd, p)
-
-		// Note: have to drop the error here as `ResolveOnce` doesn't handle 'leaf'
-		// paths (so e.g. for `echo '{"foo":123}' | ipfs dag put` we wouldn't be
-		// able to resolve `zdpu[...]/foo`)
-		if lnk == nil {
-			break
-		}
-
-		if err != nil {
-			if err == dag.ErrLinkNotFound {
-				err = ErrNoLink{Name: p[0], Node: nd.Cid()}
-			}
-			return cid.Cid{}, nil, err
-		}
-
-		if len(rest) == 0 {
-			return lnk.Cid, nil, nil
-		}
+	// resolve node before last path segment
+	nodes, err := r.resolveNodes(ctx, c, pathSelector)
+	if err != nil {
+		return cid.Cid{}, nil, err
+	}
+	if len(nodes) < 1 {
+		return cid.Cid{}, nil, fmt.Errorf("path %v did not resolve to a node", fpath)
+	}
+	parent := nodes[len(nodes)-1]
 
-		next, err := lnk.GetNode(ctx, r.DAG)
-		if err != nil {
-			return cid.Cid{}, nil, err
-		}
-		nd = next
-		p = rest
+	// find final path segment within node
+	nd, err := parent.LookupByString(p[len(p)-1])
+	if err != nil {
+		return cid.Cid{}, nil, err
 	}
 
-	if len(p) == 0 {
-		return nd.Cid(), nil, nil
+	if nd.Kind() != ipldp.Kind_Link {
+		return cid.Cid{}, nil, fmt.Errorf("path %v resolves to something other than a cid link: %v", fpath, nd)
 	}
 
-	// Confirm the path exists within the object
-	val, rest, err := nd.Resolve(p)
+	lnk, err := nd.AsLink()
 	if err != nil {
-		if err == dag.ErrLinkNotFound {
-			err = ErrNoLink{Name: p[0], Node: nd.Cid()}
-		}
 		return cid.Cid{}, nil, err
 	}
 
-	if len(rest) > 0 {
-		return cid.Cid{}, nil, errors.New("path failed to resolve fully")
-	}
-	switch val.(type) {
-	case *ipld.Link:
-		return cid.Cid{}, nil, errors.New("inconsistent ResolveOnce / nd.Resolve")
-	default:
-		return nd.Cid(), p, nil
+	clnk, ok := lnk.(cidlink.Link)
+	if !ok {
```

Comment:
eventually we may want to have a slow fallback of
`cid.ParseString(lnk.String())` to support other concrete cid types.

cc @warpfork  on thoughts on if needed / other ways to make our life remain extensible here.

Reply:
nah, honestly I don't even know if the whole `, ok` and branch seems relevant to me.

It doesn't hurt anything to ask the cast rather than assert it, but I'd say that it's legitimately a compile-time bug if a program encounters an unexpected concrete type inside this interface, and so if this wanted to panic when that fails, that would make sense to me.